api/src/main/java/com/spotify/ffwd/util/HighFrequencyDetector.java

```diff
@@ -0,0 +1,157 @@
+          b. send both with many points
+          c. check swap of high freq hashmap
+   */
+
+  @Inject
+  public HighFrequencyDetector() {
+    this.highFrequencyTriggersTS = new AtomicLong(System.currentTimeMillis());
+    this.highFrequencyRecycleLock = new AtomicBoolean(false);
+  }
+
+  public List<Metric> detectHighFrequencyMetrics(final List<Metric> metrics) {
+    List<Metric> newList = new ArrayList<>();
+
+    // Groups metrics by metric hash code and finds timestamp deltas of ordered data points.
+    // {hashcode -> -1, hashcode1 -> 10}
+    Map<Integer, Integer> groupedMetrics =
+        metrics.stream()
+            .sorted(Comparator.comparing(Metric::getTime))
+            .collect(
+                Collectors.groupingBy(
+                    Metric::hashCode,
+                    Collectors.collectingAndThen(
+                        Collectors.toList(),
+                        list -> {
+                          int size = list.size();
+                          return IntStream.range(1, size)
+                              .map(
+                                  x ->
+                                      (int)
+                                          (list.get(size - x).getTime().getTime()
+                                              - list.get(size - x - 1).getTime().getTime()))
+                              .filter(d -> (d >= 0 && d < minFrequencyMillisAllowed))
+                              .min()
+                              .orElse(-1);
+                        })));
+
+    if (dropHighFrequencyMetric) {
+      updateHighFrequencyMetricsStats(groupedMetrics);
+
+      if (highFrequencyMetrics.get().size() > 0) {
+        metrics.stream()
+            .filter(
+                metric ->
+                    highFrequencyMetrics.get().getOrDefault(metric.hashCode(), 0)
+                        < minNumberOfTriggers)
+            .forEach(newList::add);
+
+        statistics.reportHighFrequencyMetricsDropped(metrics.size() - newList.size());
+        return newList;
+      }
+    }
+
+    return metrics;
+  }
+
+  /**
+   * Updates internal map of high frequency metrics
+   *
+   * @param groupedMetrics - Grouped metrics
+   */
+  private void updateHighFrequencyMetricsStats(final Map<Integer, Integer> groupedMetrics) {
+    groupedMetrics.entrySet().stream()
+        .filter(x -> x.getValue() >= 0)
+        .forEach(
+            hashcode -> {
+              highFrequencyMetrics
+                  .get()
+                  .compute(hashcode.getKey(), (key, val) -> (val == null) ? 1 : val + 1);
+            });
+
+    statistics.reportHighFrequencyMetrics(highFrequencyMetrics.get().size());
+    swapHighFrequencyTriggersData();
+  }
+
+  /** This will reset high frequency triggers data */
+  private void swapHighFrequencyTriggersData() {
+    if (System.currentTimeMillis() - highFrequencyTriggersTS.get() > highFrequencyDataRecycleMS
+        && highFrequencyRecycleLock.compareAndExchange(false, true)) {
```

Comment:
I'm a little confused on why an AtomicBoolean is used, could this just be `synchronized(some object) { .. }` or use a [`Lock`](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/Lock.html)? seems like the highFrequencyRecycleLock's value is always reset back to `false` when this block ends - is the idea that if two threads call it at the same time, only one should enter (as opposed to a lock where both could grab the lock,  one after the other)? 

assuming a simple Lock doesn't fit the need, a comment here to explain the pattern could help

Reply:
I've updated to use `synchronized(some object) { .. }` as it should be sufficient for our case.
The idea is to restart stats.